edi-module/segment_group_reader.bal

```diff
@@ -0,0 +1,187 @@
+            // as the segment is repeatable.
+            var segments = sgContext.segmentGroup[segSchema.tag];
+            if (segments is EDISegment[]) {
+                if (segSchema.maxOccurances != -1 && segments.length() >= segSchema.maxOccurances) {
+                    return error(string `${segSchema.code} is repeatable segment with maximum limit of ${segSchema.maxOccurances}.
+                    EDI document contains more such segments than the allowed limit. Current row: ${context.rawIndex}`);
+                }
+                segments.push(segment);
+            } else if segments is () {
+                segments = [segment];
+                sgContext.segmentGroup[segSchema.tag] = segments;
+            } else {
+                return error(string `${segSchema.code} must be a segment array.`);
+            }
+        }
+    }
+
+    function placeEDISegmentGroup(EDISegmentGroup segmentGroup, EDISegGroupSchema segGroupSchema, SegmentGroupContext sgContext, EDIContext context) returns error? {
+        if segGroupSchema.maxOccurances == 1 {
+            // This is a non-repeatable mapping. So we have to compare the next segment with the next mapping.
+            log:printDebug(string `Completed reading non-repeating segment group ${printSegGroupMap(segGroupSchema)} | Current segment text: ${context.rawIndex < context.rawSegments.length() ? context.rawSegments[context.rawIndex] : "-- EOF --"}`);
+            sgContext.schemaIndex += 1;
+            sgContext.segmentGroup[segGroupSchema.tag] = segmentGroup;
+        } else {
+            // This is a repeatable mapping. So we compare the next segment also with the current mapping.
+            // i.e. we don't increment the mapping index.
+            var segmentGroups = sgContext.segmentGroup[segGroupSchema.tag];
+            if segmentGroups is EDISegmentGroup[] {
+                if segGroupSchema.maxOccurances != -1 && segmentGroups.length() >= segGroupSchema.maxOccurances {
+                    return error(string `${printSegGroupMap(segGroupSchema)} is repeatable segment group with maximum limit of ${segGroupSchema.maxOccurances}.
+                    EDI document contains more such segment groups than the allowed limit. Current row: ${context.rawIndex}`);
+                }
+                segmentGroups.push(segmentGroup);
+            } else if segmentGroups is () {
+                segmentGroups = [segmentGroup];
+                sgContext.segmentGroup[segGroupSchema.tag] = segmentGroups;
+            } else {
+                return error(string `${segGroupSchema.tag} must be a segment group array.`);
+            }
+
+        }
+    }
+
+    function validateRemainingSchemas(SegmentGroupContext sgContext) returns error? {
+        if sgContext.schemaIndex < sgContext.unitSchemas.length() - 1 {
+            int i = sgContext.schemaIndex + 1;
+            while i < sgContext.unitSchemas.length() {
```

Comment:
Can'we use a foreach statement instead with range `i ..< sgContext.unitSchemas.length()`?